File: backends/nodejs/api/algorithms.ts
import express from 'express'
import { Context, measurePerformance } from '../utils/utils'

const router = express.Router()

// router
router.get('/prime', async function (req, res) {
  const time = await measurePerformance(prime, { n: 10000000 }, false)
  res.send(time.toFixed(0))
})
router.get('/fast-fibonacci', async function (req, res) {
  const time = await measurePerformance(fastFibonacci, { n: 1000000 }, false)
  res.send(time.toFixed(0))
})
router.get('/quicksort', async function (req, res) {
  const n = 1000000
  const arr = Array(n)
  for (let i = 0; i < arr.length; i++) {
    arr[i] = n - i
  }
  const time = await measurePerformance(quicksort, { n: 10000000, array: arr }, false)
  res.send(time.toFixed(0))
})
router.get('/two-sum', async function (req, res) {
  const n = 1000000
  const arr = Array(n)
  for (let i = 0; i < arr.length; i++) {
    arr[i] = i + 1
  }
  const time = await measurePerformance(twoSum, { n: n, array: arr }, false)
  res.send(time.toFixed(0))
})

// functions

// /prime
export function prime(ctx: Context) {
  const n = ctx.n

  let p = 2
  const primes = Array(n + 1).fill(false)

  let j = n - 1

  while (p * p <= n) {
    if (!primes[p - 1]) {
      for (let i = p * p; i <= n; i += p) {
        if (!primes[i - 1]) {
          j -= 1
        }
        primes[i - 1] = true
      }
    }
    p += 1
  }
  console.log('Number of prime numbers until ' + n + ': ' + j)
}

// /fast-fibonacci
function fastFibonacci(ctx: Context) {
  const n = ctx.n
  const baseMatrix = [
    [BigInt(1), BigInt(1)],
    [BigInt(1), BigInt(0)]
  ]
  powerMatrix(baseMatrix, n)
}

function powerMatrix(matrix: bigint[][], n: number): bigint[][] {
  if (n === 1) {
    return matrix
  }
  if (n % 2) {
    return matrixMult(matrix, powerMatrix(matrix, n - 1))
  }
  const root = powerMatrix(matrix, n / 2)
  return matrixMult(root, root)
}

function matrixMult(a: bigint[][], b: bigint[][]): bigint[][] {
  const res: bigint[][] = [[], []]

  res[0][0] = a[0][0] * b[0][0] + a[0][1] * b[1][0]
  res[0][1] = a[0][0] * b[0][1] + a[0][1] * b[1][1]
  res[1][0] = a[1][0] * b[0][0] + a[1][1] * b[1][0]
  res[1][1] = a[1][0] * b[0][1] + a[1][1] * b[1][1]

  return res
}

// /quicksort
function quicksort(ctx: Context) {
  const baseArray = ctx.array as number[]

  console.log(divideArray(baseArray, 0, baseArray.length - 1)[0])
}

function divideArray(array: number[], left: number, right: number) {
  if (left >= right) {
    return array
  }

  const pivot = Math.floor((right + left) / 2)

  let temp = array[pivot]
  array[pivot] = array[right]
  array[right] = temp

  let j = left
  for (let i = left; i < right + 1; i++) {
    if (array[i] < array[right]) {
      temp = array[i]
      array[i] = array[j]
      array[j] = temp
      j++
    }
  }

  temp = array[j]
  array[j] = array[right]
  array[right] = temp

  divideArray(array, left, j - 1)
  divideArray(array, j + 1, right)

  return array
}

// /two-sum
function twoSum(ctx: Context) {
  const n = ctx.n
  const array = ctx.array as number[]
  let count = 0
  let complement
  const map = new Map<number, number>()

  for (const number of array) {
    complement = n - number
    if (map.has(complement)) {
      count += map.get(complement)!
    }

    if (map.has(number)) {
      const oldNumber = map.get(number)!
      map.set(number, oldNumber + 1)
    }
    if (!map.has(number)) {
      map.set(number, 1)
    }
  }
  console.log(count)
}

export default router

File: backends/go/api/algorithms.go
package api

import (
	"benchmarks/utils"
	"fmt"
	"math/big"
)

// /prime
func PrimeNumbers(opt *utils.Options){
	p := 2
	n := opt.N
	primes := make([]bool, n+1)

	// exclude 1
	j := n-1

	// sieve of eratosthenes
	for p*p <= n{
		if !primes[p-1]{
			for i:=p*p; i<=n; i+=p{
				if !primes[i-1]{
					j -= 1
				}
				primes[i-1] = true
			} 
		}
		p += 1
	}
	fmt.Printf("Number of prime numbers until %d: %v", n, j)
}

// /fast-fibonacci
func FastFibonacci(opt *utils.Options){
	n := opt.N
	baseMatrix := [2][2]*big.Int{
		{big.NewInt(1),big.NewInt(1)},
		{big.NewInt(1),big.NewInt(0)},
	}
	powerMatrix(baseMatrix, n)
}

func powerMatrix(matrix [2][2]*big.Int, n int) [2][2]*big.Int{
	if n==1{
		return matrix
	}
	if n%2==1{
		return matrixMult(matrix, powerMatrix(matrix, n-1))
	}
	root := powerMatrix(matrix, n/2)
	return matrixMult(root, root)
}

func matrixMult(a [2][2]*big.Int, b [2][2]*big.Int) [2][2]*big.Int{
	var res [2][2]*big.Int
	for i := range res {
		for j := range res[i] {
			res[i][j] = new(big.Int)
		}
	}
	res[0][0].Mul(a[0][0], b[0][0]).Add(res[0][0], new(big.Int).Mul(a[0][1], b[1][0]))
	res[0][1].Mul(a[0][0], b[0][1]).Add(res[0][1], new(big.Int).Mul(a[0][1], b[1][1]))
	res[1][0].Mul(a[1][0], b[0][0]).Add(res[1][0], new(big.Int).Mul(a[1][1], b[1][0]))
	res[1][1].Mul(a[1][0], b[0][1]).Add(res[1][1], new(big.Int).Mul(a[1][1], b[1][1]))

	return res
}

// /quicksort
func Quicksort(opt *utils.Options){
	baseArray := opt.Array
	
	fmt.Println(divideArray(baseArray)[0])
}

func divideArray(array []int) []int{
	length := len(array)
	if length < 2{
		return array
	}
	// the fact that go rounds integers down if you get a float by dividing an int is really interesting
	pivot := length/2
	left := 0
	
	array[length-1], array[pivot] = array[pivot], array[length-1]

	for i := range array {
		if array[i] < array[length-1] {
			array[i], array[left] = array[left], array[i]
			left++
		}
	}
	array[left], array[length-1] = array[length-1], array[left]

	divideArray(array[:left])
	divideArray(array[left+1:])

	return array
}

// /two-sum
func TwoSum(opt *utils.Options){
	n := opt.N
	hashmap := make(map[int]int, n)
	arr := opt.Array
	count := 0

	for _, number := range arr{
		complement := n - number
		count += hashmap[complement]
		hashmap[number] += 1
	}
	fmt.Println(count)
}